pkg/auth/oauth/registry/grantchecker.go

```diff
@@ -30,12 +30,22 @@ func (c *ClientAuthorizationGrantChecker) HasAuthorizedClient(user user.Info, gr
 	if err != nil {
 		return false, err
 	}
-	if len(authorization.UserUID) != 0 && authorization.UserUID != user.GetUID() {
-		return false, fmt.Errorf("user %s UID %s does not match stored client authorization value for UID %s", user.GetName(), user.GetUID(), authorization.UserUID)
+
+	// Validation prevents authorization.UserUID from being empty (and always has).
+	// However, user.GetUID() is empty during impersonation, meaning this flow does not work for impersonation.
+	// This is fine because no OAuth / grant flow works with impersonation in general.
+	if user.GetUID() != authorization.UserUID {
+		glog.Infof("%#v does not match stored client authorization %#v, attempting to delete stale authorization", user, authorization)
+		if err := c.client.Delete(id, nil); err != nil && !errors.IsNotFound(err) {
```

Comment:
add delete options with a precondition that ensures the authorization we delete *this* oauthclientauthorization:

NewPreconditionDeleteOptions(authorization.UID)

determine how to handle an error from deletion (tolerate notfound, refetch/check on conflict?, fail on other errors)